### PR TITLE
Correct unconditionally bridge for a number of Foundation types

### PR DESCRIPTION
**Pull request by phausler**
Foundation incorrectly implemented the unconditional bridge methods in a number of cases that could represent a reasonable fallback for undecorated or improperly decorated APIs. This changes Data to return a Data with zero bytes, AffineTransform to return identity, CharacterSet to return an empty CharacterSet, DateComponents to return a structure with no values, Decimal to return 0, IndexSet to return an empty IndexSet, IndexPath to return an empty IndexPath, Notification to return an "empty" notification (note the name is just the empty string; which is not suggested to subscribe to or post), and URLComponents to return a structure with no values when the value passed was detected as a nil value when it was not expected. This will help alleviate crashes when interoperating with APIs that have not yet been annotated with nullability (it is strongly suggested not just for the health and fitness of APIs to annotate for just swift but objective-c in general becomes more robust when nullability is marked 😉). 

This addresses the following issues:
rdar://problem/30398990
rdar://problem/30384882
rdar://problem/30384851
rdar://problem/30384806
rdar://problem/30384727
rdar://problem/30384692

### CODE REVIEW

**Comment by phausler** [comment]
@swift-ci please test

**Comment by jrose-apple** [comment]
A nitpick: this doesn't affect APIs that haven't been annotated for nullability, only those that have been *incorrectly* annotated.

**Comment by phausler** [comment]
Yea i should have been more specific. Out of the Foundation types these are the ones that directly have a sensible empty representation, I have a change for the NSValue ones but i need to test it more

File: stdlib/public/SDK/Foundation/DateComponents.swift [hunk 1]
```diff
@@ -339,9 +339,8 @@ extension DateComponents : _ObjectiveCBridgeable {
     }
 
     public static func _unconditionallyBridgeFromObjectiveC(_ source: NSDateComponents?) -> DateComponents {
-        var result: DateComponents?
-        _forceBridgeFromObjectiveC(source!, result: &result)
-        return result!
+        guard let src = source else { return DateComponents() }
+        return DateComponents(reference: src)
```
Comment (michael-lehew) [inline stdlib/public/SDK/Foundation/DateComponents.swift]: unrelated, but its a shame how many names we have for the first paraemeter (_bridged / referencing / reference / etc.)

**Review by michael-lehew — APPROVED** [review]
I think this looks good to me!

File: stdlib/public/SDK/Foundation/DateComponents.swift (hunk 1 above)
Comment (phausler) [inline stdlib/public/SDK/Foundation/DateComponents.swift]: most of those are internal; I think it would be interesting to perhaps make all bridgeable types have a public init method that takes a reference and a counterpart to the reference types that take the value (however that may seem too C++ to some...)

**Comment by phausler** [comment]
Tony requested behavior tests for these; holding off on merge until I have those ready

File: stdlib/public/SDK/Foundation/AffineTransform.swift [hunk 2]
```diff
@@ -314,9 +314,8 @@ extension AffineTransform : _ObjectiveCBridgeable {
     }
 
     public static func _unconditionallyBridgeFromObjectiveC(_ x: NSAffineTransform?) -> AffineTransform {
-        var result: AffineTransform?
-        _forceBridgeFromObjectiveC(x!, result: &result)
-        return result!
+        guard let src = x else { return  AffineTransform.identity }
```
Comment (xwu) [inline stdlib/public/SDK/Foundation/AffineTransform.swift]: World's smallest drive-by nit: extra space after `return` 😛

**Comment by phausler** [comment]
@swift-ci test

**Comment by phausler** [comment]
@swift-ci test

**Comment by phausler** [comment]
@swift-ci test

**Comment by phausler** [comment]
@swift-ci test

**Comment by phausler** [comment]
The failure here is not part of this commit: https://ci.swift.org/job/swift-PR-osx/6885/testReport/junit/Swift-Unit/runtime_SwiftRuntimeTests/MetadataTest_getExistentialMetadata/

@swift-ci test

**Comment by phausler** [comment]
@swift-ci please test and merge

**Comment by phausler** [comment]
```
/Users/buildnode/jenkins/workspace/swift-PR-osx/branch-master/swift/test/IRGen/indirect_argument.sil:18:14: error: expected string not found in input
16:15:02 // CHECK-32: call swiftcc void @huge_method(%T17indirect_argument4HugeV* noalias nocapture swiftself dereferenceable({{.*}})
```

I am quite certain this does not change the IRGen side of things

**Comment by phausler** [comment]
@swift-ci please test

**Comment by phausler** [comment]
```
Waiting for the completion of Swift Pull Request Test Linux
Swift Pull Request Test Linux #7156 completed. Result was FAILURE
Waiting for the completion of Swift Pull Request Test OS X
Swift Pull Request Test OS X #7199 completed. Result was FAILURE
```

swift-ci sounds really negative about my changes,
 @shahmishal is there something more here that I am missing to find where the error is?

**Comment by phausler** [comment]
lets try it again for good measure

@swift-ci please test and merge
